safenode/src/protocol/types/wallet/local_store.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{CreatedDbc, DepositWallet, Result, SendClient, SendWallet};
+
+use crate::protocol::types::transfers::Outputs as TransferDetails;
+
+use sn_dbc::{Dbc, DbcId, DbcIdSource, MainKey, PublicAddress, Token};
+
+use async_trait::async_trait;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// A wallet that can send tokens to other addresses.
+pub struct LocalSendWallet<C: SendClient> {
+    client: C,
+    wallet: LocalDepositWallet,
+}
+
+/// A wallet that can only receive tokens.
+pub struct LocalDepositWallet {
+    /// The secret key with which we can access
+    /// all the tokens in the available_dbcs.
+    key: MainKey,
+    /// The current balance of the wallet.
+    balance: Token,
+    /// These are dbcs we've owned, that have been
+    /// spent when sending tokens to other addresses.
+    spent_dbcs: BTreeMap<DbcId, Dbc>,
+    /// These are the dbcs we own that are not yet spent.
+    available_dbcs: BTreeMap<DbcId, Dbc>,
+    /// These are the dbcs we've created by
+    /// sending tokens to other addresses.
+    output_history: Vec<CreatedDbc>,
+    // /// The path to the wallet file.
+    // local_store: std::path::PathBuf,
+}
+
+impl DepositWallet for LocalDepositWallet {
+    fn new(key: MainKey) -> Self {
+        Self {
+            key,
+            balance: Token::zero(),
+            spent_dbcs: BTreeMap::new(),
+            available_dbcs: BTreeMap::new(),
+            output_history: vec![],
+        }
+    }
+
+    // /// Loads a serialized wallet from a path.
+    // fn load_from(path: &Path) -> Self {
+    //     Self {
+    //     }
+    // }
+
+    fn address(&self) -> PublicAddress {
+        self.key.public_address()
+    }
+
+    fn balance(&self) -> Token {
+        self.balance
+    }
+
+    fn new_dbc_address(&self) -> DbcIdSource {
+        self.key.random_dbc_id_src(&mut rand::thread_rng())
+    }
+
+    fn deposit(&mut self, dbcs: Vec<Dbc>) {
```

Comment:
Perhaps it's an idea to make a method that deposits a single DBC? This could return `Option<Dbc>`, moving the `Dbc` back to the caller if the DBC is not deposited. Then adjust the balance only by adding up the amount of the DBC just added.

Feel free to ignore, just an idea.